polyply/src/gen_seq.py

```diff
@@ -0,0 +1,254 @@
+            self.residues.append(name)
+            self.weights.append(float(prob))
+
+    def gen_graph(self, seed=None):
+        """
+        Generate a graph from the defnitions stored in this
+        instance.
+        """
+        graph = _branched_graph("dum", self.bfact, self.levels)
+        graph = _random_replace_nodes(graph, self.residues, self.weights)
+        return graph
+
+
+def _add_edges(graph, edges, idx, jdx):
+    """
+    Add edges to a graph using the edge format
+    'node_1,node_2', where the node keys refer
+    to the node relative to the nodes with the same
+    attribute `seqid`. The sequence ids are provided by
+    `idx` and `jdx`.
+
+    Parameters
+    ----------
+    graph: `:class:networkx.Graph`
+    edges:  list[str]
+        str has the format `node_keyA,node_keyB-node_keyC,node_keyD`
+    idx: int
+    jdx: int
+
+    Returns
+    --------
+    `:class:networkx.Graph`
+    """
+    for edge in edges.split(","):
+        node_idx, node_jdx = edge.split("-")
+        idx_nodes = find_atoms(graph, "seqid", idx)
+        jdx_nodes = find_atoms(graph, "seqid", jdx)
+
+        if len(idx_nodes) == 0:
+            msg = ("Trying to add connect between block with seqid {} and block with"
+                   "seqid {}. However, cannot find block with seqid {}.")
+            raise IOError(msg.format(idx, jdx, idx))
+        elif len(jdx_nodes) == 0:
+            msg = ("Trying to add connect between block with seqid {} and block with"
+                   "seqid {}. However, cannot find block with seqid {}.")
+            raise IOError(msg.format(idx, jdx, jdx))
+
+        try:
+            node_i = idx_nodes[int(node_idx)]
+        except IndexError:
+            msg = ("Trying to add connect between block with seqid {} and block with"
+                   "seqid {}. However, cannot find resid {} in block with seqid {}.")
+            raise IOError(msg.format(idx, jdx, node_idx, idx))
+
+        try:
+            node_j = jdx_nodes[int(node_jdx)]
+        except IndexError:
+            msg = ("Trying to add connect between block with seqid {} and block with"
+                   "seqid {}. However, cannot find resid {} in block with seqid {}.")
+            raise IOError(msg.format(idx, jdx, node_idx, idx))
+
+        graph.add_edge(node_i, node_j)
+
+    return graph
+
+
+def generate_seq_graph(sequence, macros, connects):
+    """
+    Given a sequence definition in terms of blocks in macros
+    generate a graph and add the edges provided in connects.
+
+    Parameters:
+    ----------
+    sequence: list
+      a list of blocks defining a sequence
+    macros: dict[str, networkx.graph]
+      a dictionary of graphs defining a macro block
+    connects: list[str]
+      a list of connects
+
+    Returns:
+    -------
+    `:class:networkx.graph`
+    """
+    seq_graph = nx.Graph()
+    for idx, macro_name in enumerate(sequence):
+
+        # we need this because from_file macros
+        # are just graphs
```

Comment:
Does that mean that from_file macros need more/better/different processing, such that a macro is a macro?
Or do the ways of generating macros that are not from_file need more/better/different processing?

Reply:
from file macros are already ready to go graphs, the others are strings, which need to be interpreted to make a graph

Reply:
I strongly advocate turning all "macro"s into graphs before this function then.

Reply:
This doesn't work, because the MacroString object needs to be run when the graph is added so that each time an random element is added it is actually random and we do not just keep copying a single random fragment.

Reply:
We can make a dummy class though for the file macros?? But I don't see how that makes the code much clearer.

Reply:
That does complicate matters. I indeed think that making a FileMacro class with a `gen_graph` method is probably the cleanest way. Depending on your mood you either make it fetch the Block from the FF when that is called, or you return the cached answer. I would do the first, since that simplifies https://github.com/marrink-lab/polyply_1.0/pull/53/files/172907e5ebd7f6cf84e700eddc5cda890b847338#diff-06a240204f94f8580996ce9818dd02d0R239-R241 a little